### PR TITLE
Add country page

### PR DESCRIPTION
**Pull request by nbaldy**
There is now a functioning Country page. 
Country page can be accessed at [http://localhost:8080/country/](url)[countryCode]
Country codes are the ISO 3166 Country Codes. 

Fixes issue #60 

Need to go in and cleanup user code. I don't think it's worth doing in this PR since this is a necessary to get other functionality working. I will file an issue and get to that soon

### CODE REVIEW

File: src/main/java/com/google/codeu/data/Country.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
+package com.google.codeu.data;
+
+public class Country {
```
Comment (gjchong25) [inline src/main/java/com/google/codeu/data/Country.java:3]: This looks good so far- just out of curiosity, what would the country code be used for? Or is it just an additional piece of information? Either is fine, just want to make sure each part of the code is necessary and no fields go unused.

File: src/main/java/com/google/codeu/servlets/MessageServlet.java [hunk 2]
```diff
@@ -96,51 +84,21 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
...
-    String replacement = "<img src=\"$1\" />";
-    String textWithImagesReplaced = text.replaceAll(regex, replacement);
-
-    float sentimentScore = this.getSentimentScore(text);
+    String country = request.getParameter("country");
 
-    // check if user is valid
-    if (datastore.getUser(recipient) == null) {
-      response.sendRedirect("/users/" + user);
+    // Redirect to home on invalid country
```
Comment (gjchong25) [inline src/main/java/com/google/codeu/servlets/MessageServlet.java:91]: So could you clarify what you mean by "invalid country"? Are we not hardcoding the country data from a csv? Because for users, they were only valid when they logged in and had an aboutMe, what would an invalid country entail?

File: src/main/java/com/google/codeu/data/Country.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+  private String code;
+  private String name;
+  private double lat;
+  private double lng;
+
+  public Country(String code, String name, double lat, double lng) {
```
Comment (gjchong25) [inline src/main/java/com/google/codeu/data/Country.java:9]: I think this is still up for discussion, but how do we plan on tying in the subreddits to the Country class? The Country page should display the subreddits and several of the most recent posts, so maybe we can add a field to the Country class that keeps track of these channels somehow.

File: src/main/java/com/google/codeu/data/Country.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
+package com.google.codeu.data;
```
Comment (gjchong25) [inline src/main/java/com/google/codeu/data/Country.java:1]: Overall, I think the code looks pretty good. It might be easier to break your code up into multiple pull requests rather than doing it all in one next time. We might consider how a Country class might be similar/different to a Subchannel or subreddit class. Also, do we want users to be able to post directly in the Country class or have them find separate subreddits in the country page and then post on those respective channels instead? I think we should keep discussing with as a team how we'd like to handle cases like these.

File: src/main/java/com/google/codeu/data/Country.java (hunk 1 above)
Comment (nbaldy) [reply src/main/java/com/google/codeu/data/Country.java:3]: So, some countries are super long "The democratic republic of the Congo". I forsee that possibly causing some issues down the eline, so for now I have everything using the country codes. I do forsee a problem with compatibility once users start being able to add locations... and I'm not really sure how to deal with that quite yet. But right now, I have the country codes, and we can switch from one to another very easily if we choose to use the country names instead (if we address the spacing issues) - I guess that would be as simple as replacing spaces with nothing or underscores... I don't know. 

I partially just wanted to push this code out so that others can build on top of it, so I'm somewhat deferring these decisions until we get a little farther, because for our MVP, it doesn't matter. If it comes back to bite us on level 2, it should hopefully be a quick fix

File: src/main/java/com/google/codeu/servlets/MessageServlet.java (hunk 2 above)
Comment (nbaldy) [reply src/main/java/com/google/codeu/servlets/MessageServlet.java:91]: So, if the user somehow tried to post a message to the page of a country that doesn't exist. As the code stands right now, I don't think this is possible (Because non-existent countries don't have pages, and attempts to access the page through localhost:8080/country/[doesn't exist] are redirected to the invalid-country error page, unless the user would somehow alter the page scripts themselves. It may be an extra check, as I'm not sure that the user could ever change that form, but it's the same thing as we had of checking if users tried to post something to a non-existant user page

Long story short, yes, we do generate the country pages; but they are stored similarly to users in datastore. Therefore, we can easily check if a user is trying to send a message to a country that they doesn't exist by checking if we have data on this country

File: src/main/java/com/google/codeu/data/Country.java (hunk 3 above)
Comment (nbaldy) [reply src/main/java/com/google/codeu/data/Country.java:9]: I think that should be done soon, but also I think it's out of the scope of this particular PR. 

My goal with this PR was to get something to build off of - the more features and functionality I am trying to implement, the longer it will take; I see the country pages as being vital to much of the functionality of most other things, so I want to get this completed, and wecan add this display functionality soon

File: src/main/java/com/google/codeu/data/Country.java (hunk 4 above)
Comment (nbaldy) [reply src/main/java/com/google/codeu/data/Country.java:1]: I do like having small PRs, but the problem here is that in creating a new datastore class (so that countries exist) I have to create an awful lot of files and change some things around. Most of the changes here need to all be implemented together or the code breaks.  I probably could have held off on the "messages can be posted to country pages" bit, but I hadn't thought of it and also it was a relatively minor change that may have been more confusing to make down the line. 

/country/[country]/[subchannel] might be a good layout to have - possibly we can use the country servlet to do checks that country exists, then have different functions for loading different subchannels. As for messages, we'd have to add a "channel" data field when storing and  processing them

I know that we pretty much decided we will not post directly to the Country page, but to me, changing message functionality from to-user to to-country was a nice half-step, where it's easy to build off of and change later, as people are actually building the subchannels they can change the messaging capability as needed.

File: src/main/java/com/google/codeu/data/Datastore.java [hunk 5]
```diff
@@ -134,10 +143,26 @@ public User getUser(String email) {
...
       return null;
     }
-
     // Return aboutMe
     String aboutMe = (String) userEntity.getProperty("aboutMe");
     User user = new User((String) userEntity.getProperty("email"), aboutMe);
     return user;
   }
+  /** Returns true if country is stored in datastore, false o.w. */
+  public Country getCountry(String countryCode) {
```
Comment (drewbern) [inline src/main/java/com/google/codeu/data/Datastore.java:155]: If a country exists, should this return true (as in the comment), or should it return the country object?

File: src/main/java/com/google/codeu/servlets/CountryServlet.java [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+import java.io.IOException;
+import java.util.List;
+import java.util.Scanner;
+import javax.servlet.ServletException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+@WebServlet("/country/*")
```
Comment (drewbern) [inline src/main/java/com/google/codeu/servlets/CountryServlet.java:17]: Is this servlet also going to handle requests to subcategories for the country? For example, /country/Canada/Food ?

**Review by drewbern — COMMENTED** [review]
LGTM! Thanks for doing all of this!

File: src/main/java/com/google/codeu/data/Datastore.java (hunk 5 above)
Comment (nbaldy) [reply src/main/java/com/google/codeu/data/Datastore.java:155]: Good catch, I originally had it as a isCountryValid message and forgot to change the comment. I'll do that :D

File: src/main/java/com/google/codeu/servlets/CountryServlet.java (hunk 6 above)
Comment (nbaldy) [reply src/main/java/com/google/codeu/servlets/CountryServlet.java:17]: I think it would be good to do so, eventually. Right now, it doesn't have separate processing for different urls, but I do think that would be a good way to set up that url and use wildcarding

**Review by drewbern — APPROVED** [review]

**Review by ArgaAlan — APPROVED** [review]
LGTM
